lib/MusicBrainz/Server/Data/Utils.pm

```diff
@@ -364,6 +364,8 @@ sub trim_comment {
 sub trim_text {
     my $t = shift;
 
+    $t = NFC($t);
+    $t = remove_invalid_characters($t);
```

Comment:
The other place we use this has a comment:
```
    # Before removing invalid characters, convert space control characters
    # into U+0020 (or else they'll be removed).

```
Is this a worry here?

Reply:
That comment is not accurate enough, it does not give any example of such character. For example, TAB character is not removed, so I’m not sure which “space control characters” would actually benefit from that.

The other place collapses whitespaces, which we cannot do here because of possible indentation.

It seems neither needed nor appropriate to replace unwanted control characters such as bell with whitespace.

Reply:
It's talking about any character removed by `remove_invalid_characters` -- there's an explicit list in there (which includes TAB, LF, CR).

Were you perhaps testing this in the release editor? Because that goes through `WS::js::Edit`, not `Form::Annotation`. If you test on any `/edit_annotation` page, this patch will indeed cause it to remove TAB, LF, CR, etc.

(This also suggests we'll have to call `trim_text` where annotations are processed in `WS::js::Edit`.)

Reply:
> there's an explicit list in there (which includes TAB, LF, CR).

This was not clear because [`remove_invalid_characters`](https://github.com/metabrainz/musicbrainz-server/blob/v-2021-03-01/lib/MusicBrainz/Server/Data/Utils.pm#L398-L411) currently starts by explicitly NOT removing TAB, LF, CR with `s/[^\x09\x0A\x0D…]//gor` but it ends with removing them with `s/[…\p{Cc}…]//g`.

This function will be split using commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/2bc4060b8f1abd524b95f2d035546273ee0ff119.

> we'll have to call `trim_text` where annotations are processed in `WS::js::Edit`.

Fixed in commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/dbea2f2446aad6aebf8e388d6b4025fe3c268247.

It’s a bit an issue that data processors in `WS::js::Edit` are not systematically synced with their `Form` counterparts. What can we do about that, commenting code with reminders?

Reply:
Comments are probably the best we can do for now until more significant improvements can be made. If I could go back, I'd probably make /ws/js/edit submit everything through FormHandler and just serialize the field errors it returns.

Reply:
Reported as ticket MBS-11428 and added comments through commit https://github.com/metabrainz/musicbrainz-server/pull/1944/commits/aee75730894e3c0bc096b5f0603b51a881afca31.